### PR TITLE
Allow gerrit reporter callbacks to be deferred

### PR DESCRIPTION
**Pull request by ryansydnor**
This addresses http://trac.buildbot.net/ticket/3522

### CODE REVIEW

File: master/buildbot/reporters/gerrit.py [hunk 1]
```diff
@@ -334,8 +335,12 @@ def getBuildInfo(build):
                         }
             buildInfoList = sorted([getBuildInfo(build) for build in builds], key=lambda bi: bi['name'])
 
-            result = _handleLegacyResult(self.summaryCB(buildInfoList, Results[buildset['results']],
-                                                        self.master, self.summaryArg))
+            result = yield self.summaryCB(buildInfoList,
```
Comment (tardyp) [inline master/buildbot/reporters/gerrit.py]: I think we should do this for all callbacks

Comment (ryansydnor) [inline master/buildbot/reporters/gerrit.py]: @tardyp - agree! Do you want to do this as part of this PR? Or use this as the example for future PRs?

Comment (tardyp) [inline master/buildbot/reporters/gerrit.py]: I think we should keep coherency, and do everything at once (as it is not huge change)

Comment (ryansydnor) [inline master/buildbot/reporters/gerrit.py]: Okay. I'll work on that tonight for the gerrit reporter.

Comment (ryansydnor) [inline master/buildbot/reporters/gerrit.py]: @tardyp - done!

File: master/buildbot/test/unit/test_reporter_gerrit.py [hunk 2]
```diff
@@ -70,6 +85,29 @@ def testSummaryCB(buildInfoList, results, status, arg):
...
+            failure = True
+
+    if failure:
+        verified = -1
+    elif success:
+        verified = 1
+    else:
+        verified = 0
+
+    defer.returnValue(makeReviewResult(str(buildInfoList),
```
Comment (tardyp) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: You should rather do:

    result = yield makeReviewResult(...)
    self.expect...(results[..], foo)

File: master/buildbot/test/unit/test_reporter_gerrit.py [hunk 3]
```diff
@@ -44,11 +44,26 @@ def testReviewCB(builderName, build, result, status, arg):
...
+
+
 def testStartCB(builderName, build, arg):
     return makeReviewResult(str({'name': builderName}),
                             (GERRIT_LABEL_REVIEWED, 0))
 
 
+@defer.inlineCallbacks
+def testStartCBDeferred(builderName, build, arg):
+    yield
```
Comment (tardyp) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: This does not make sense to yield nothing in a inlineCallback.

Please look at other tests to see how we usually do.

File: master/buildbot/test/unit/test_reporter_gerrit.py (hunk 2 above)
Comment (ryansydnor) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: Done.

File: master/buildbot/test/unit/test_reporter_gerrit.py (hunk 3 above)
Comment (ryansydnor) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: @tardyp - thanks for bearing with me while I learn the twisted framework. 

I thought yielding nothing was "good enough" for a test - it showed we were capable of calling a deferred if necessary. Your proposal makes much more sense, though! Fixed.

File: master/buildbot/test/unit/test_reporter_gerrit.py [hunk 4]
```diff
@@ -70,6 +85,29 @@ def testSummaryCB(buildInfoList, results, status, arg):
...
+    if failure:
+        verified = -1
+    elif success:
+        verified = 1
+    else:
+        verified = 0
+
+    result = yield makeReviewResult(str(buildInfoList),
+                                    (GERRIT_LABEL_VERIFIED, verified))
+    defer.returnValue(result)
```
Comment (tardyp) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: don't you want to at least make sure the result variable is coherent with what you expect?

You can use one of those:
https://twistedmatrix.com/documents/14.0.2/api/twisted.trial._synctest._Assertions.html

Comment (ryansydnor) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: That already happens on line 211.

Basically, I'm testing that the result is exactly the same as calling the summary CB in a non-deferred way.

Comment (tardyp) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: Ah I thoughts those callbacks where tests.

There is a convention that says that methods starting with test are tests.
Please rename those function, like testedSummaryCBDeferred

Comment (ryansydnor) [inline master/buildbot/test/unit/test_reporter_gerrit.py]: Good point. Fixed.

**Comment by tardyp** [comment]
👍

**Comment by ryansydnor** [comment]
@tardyp - waiting on anyone else to review?

**Comment by tardyp** [comment]
I let some time for other reviewers to catch up

**Comment by sa2ajj** [comment]
LGTM

**Pull request merged**